ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java

```diff
@@ -0,0 +1,154 @@
+import com.microsoft.azure.storage.CloudStorageAccount;
+import com.microsoft.azure.storage.OperationContext;
+import com.microsoft.azure.storage.StorageException;
+import com.microsoft.azure.storage.blob.BlobContainerPublicAccessType;
+import com.microsoft.azure.storage.blob.BlobRequestOptions;
+import com.microsoft.azure.storage.blob.CloudBlobClient;
+import com.microsoft.azure.storage.blob.CloudBlobContainer;
+import com.microsoft.azure.storage.blob.CloudBlockBlob;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Implementation of {@link CloudDestination} that interacts with Azure Blob Storage service.
+ */
+class AzureCloudDestination implements CloudDestination {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AzureCloudDestination.class);
+  private final CloudStorageAccount azureAccount;
+  private final CloudBlobClient azureBlobClient;
+
+  /**
+   * Construct an Azure cloud destination from a container's replication config.
+   * @param configSpec the config spec to use.
+   * @throws CloudStorageException if the destination could not be created.
+   */
+  AzureCloudDestination(String configSpec) throws CloudStorageException {
+    this(configSpec, null);
+  }
+
+  /**
+   * Construct an Azure cloud destination from a container's replication config and a {@link CloudStorageAccount} instance.
+   * @param configSpec the config spec to use.
+   * @param azureAccount the {@link CloudStorageAccount} to use.
+   * @throws CloudStorageException if the destination could not be created.
+   */
+  AzureCloudDestination(String configSpec, CloudStorageAccount azureAccount) throws CloudStorageException {
+    try {
+      if (azureAccount == null) {
+        azureAccount = CloudStorageAccount.parse(configSpec);
+      }
+      this.azureAccount = azureAccount;
+
+      // Create a blob client to interact with Blob storage
+      azureBlobClient = azureAccount.createCloudBlobClient();
+      LOGGER.info("Created Azure destination");
+    } catch (Exception e) {
+      throw new CloudStorageException("Failed to create AzureCloudDestination", e);
+    }
+  }
+
+  @Override
+  public boolean uploadBlob(BlobId blobId, long blobSize, InputStream blobInputStream) throws CloudStorageException {
+
+    Objects.requireNonNull(blobId, "BlobId cannot be null");
+    Objects.requireNonNull(blobInputStream, "Input stream cannot be null");
+
+    BlobRequestOptions options = null; // may want to set BlobEncryptionPolicy here
+    OperationContext opContext = null;
+    try {
+      CloudBlobContainer azureContainer = getContainer(blobId, true);
+      CloudBlockBlob azureBlob = azureContainer.getBlockBlobReference(blobId.getID());
+
+      if (azureBlob.exists()) {
+        LOGGER.debug("Skipping upload of blob {} as it already exists in Azure container {}.", blobId,
+            azureContainer.getName());
+        return false;
```

Comment:
should this be an exception? I believe that's what the `BlobStore` does

Reply:
The upload method in CloudDestination is meant to be idempotent.
I could check the return code in CloudBlobStore and throw a StoreException, but this doesn't seem like a good idea because each VCR will pull from all local replicas.  There is a chance that between checkMissingKeys() and the write operation, the blob could have been written by another replica thread.  I prefer to play it safe, even if the logic differs from the existing BlobStore.